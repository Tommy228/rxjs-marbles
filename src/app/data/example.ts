import { Observable, SchedulerLike } from 'rxjs';

/**
 * Represents an example input that can be either of type TInput or a cancellation.
 * @template TInput - The type of the input.
 */
export type ExampleInput<TInput> =
  | {
      t: number;
      x: TInput;
      c?: undefined;
    }
  | {
      t: number;
      c: true;
    };

/**
 * Unpacks an array to keep the generic type only.
 * @example UnpackedArray<Array<number>> = number
 */
export type UnpackedArray<T> = T extends Array<infer U> ? U : T;

/**
 * Takes an array of types, and boxifies them into an array of observables.
 * @example BoxifiedToObservable<[string, number]> = [Observable<string>, Observable<number>]
 */
export type BoxifiedToObservable<T> = {
  [P in keyof T]: Observable<UnpackedArray<T[P]>>;
};

/**
 * Defines the "apply" (i.e. transform) function of an example.
 * @example ExampleApplyFn<[string, boolean], number> =
 *  (inputs: [Observable<string>, Observable<boolean>], scheduler: SchedulerLike) => Observable<number>
 */
export type ExampleApplyFn<TInputs extends unknown[], TOutput> = (
  inputs: [...BoxifiedToObservable<TInputs>],
  scheduler: SchedulerLike,
) => Observable<TOutput>;

/**
 * Takes an array of types, and boxifies them into an array of example inputs
 * @example BoxifiedToExampleInput<[string, number]> = [ExampleInput<string>[], ExampleInput<number>[]]
 */
export type BoxifiedToExampleInput<T> = {
  [P in keyof T]: ExampleInput<UnpackedArray<T[P]>>[];
};

export interface Example<TInputs extends unknown[], TOutput> {
  name: string;
  description: string;
  linkToDocumentation: string;
  label: string;
  inputs?: [...BoxifiedToExampleInput<TInputs>];
  completionFrame?: number;
  apply: ExampleApplyFn<TInputs, TOutput>;
}

export type ExampleFactory = <R>(
  cb: <TInputs extends unknown[], TOutput>(
    example: Example<TInputs, TOutput>,
  ) => R,
) => R;

export const createExample =
  <TInputs extends unknown[], TOutput>(
    d: Example<TInputs, TOutput>,
  ): ExampleFactory =>
  (cb) =>
    cb(d);
