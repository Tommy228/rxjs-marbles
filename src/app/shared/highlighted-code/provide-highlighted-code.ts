import { APP_INITIALIZER, Provider } from '@angular/core';
import { provideHighlightOptions } from 'ngx-highlightjs';
import { HighlightedCodeDarkModeHandler } from './services/highlighted-code-dark-mode-handler.service';

export const HighlightJsThemes = {
  light: 'highlightjs/styles/github.css',
  dark: 'highlightjs/styles/github-dark.css',
};

export const provideHighlightedCode = (): Provider[] => [
  provideHighlightOptions({
    coreLibraryLoader: () => import('highlight.js/lib/core'),
    languages: {
      javascript: () => import('highlight.js/lib/languages/javascript'),
    },
    themePath: HighlightJsThemes.light,
  }),
  {
    provide: APP_INITIALIZER,
    multi: true,
    deps: [HighlightedCodeDarkModeHandler],
    useFactory: (handler: HighlightedCodeDarkModeHandler) => () =>
      handler.autoSwitchDarkMode(),
  },
];
