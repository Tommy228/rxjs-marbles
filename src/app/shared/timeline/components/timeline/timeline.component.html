<div
  class="timeline"
  (resized)="onResize()"
  data-timeline
>
  @if (hasName) {
    <div class="timeline__name">
      {{ name }}
    </div>
  }

  <div
    class="timeline__line timeline-line"
    #timelineLine
    data-timeline-line
  >
    <svg
      class="timeline-line-drawing"
      viewBox="0 0 50 50"
      preserveAspectRatio="none"
    >
      <line
        class="timeline-line-drawing__line"
        x1="0"
        y1="25"
        x2="50"
        y2="25"
        stroke-width="1"
      ></line>
    </svg>

    @for (element of elements ?? []; track element) {
      @if (element | guardType: isValue; as valueElement) {
        <app-timeline-value
          class="timeline-value"
          [ngClass]="'timeline-value--' + getElementColor(valueElement)"
          [value]="valueElement.value"
          [isDraggable]="isDraggable ?? true"
          cdkDrag
          [cdkDragDisabled]="!isDraggable"
          [cdkDragLockAxis]="'x'"
          [cdkDragBoundary]="timelineLine"
          (cdkDragMoved)="onDragMoved($event, valueElement)"
          [cdkDragFreeDragPosition]="getForcedDragPosition(element)!"
        />
      }

      @if (element | guardType: isCompletion; as completionElement) {
        <app-timeline-completion
          class="timeline-completion"
          cdkDrag
          [cdkDragDisabled]="!isDraggable"
          [cdkDragLockAxis]="'x'"
          [cdkDragBoundary]="timelineLine"
          (cdkDragMoved)="onDragMoved($event, completionElement)"
          [cdkDragFreeDragPosition]="getForcedDragPosition(element)!"
        />
      }

      @if (element | guardType: isError; as errorElement) {
        <app-timeline-error
          class="timeline-error"
          cdkDrag
          [cdkDragDisabled]="!isDraggable"
          [cdkDragLockAxis]="'x'"
          [cdkDragBoundary]="timelineLine"
          (cdkDragMoved)="onDragMoved($event, errorElement)"
          [cdkDragFreeDragPosition]="getForcedDragPosition(element)!"
        />
      }
    }
  </div>

  <svg
    class="timeline-arrow-drawing"
    viewBox="0 0 40 50"
    preserveAspectRatio="none"
  >
    <line
      class="timeline-arrow-drawing__line"
      x1="0"
      y1="25"
      x2="20"
      y2="25"
      stroke-width="1"
    ></line>
    <polygon
      class="timeline-arrow-drawing__polygon"
      points="20,15 40,25 20,35"
      fill="black"
    ></polygon>
  </svg>
</div>
