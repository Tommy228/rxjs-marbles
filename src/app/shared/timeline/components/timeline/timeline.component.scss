@use 'sass-rem' as rem;
@use '../../../../../sass/themes' as themes;
@use '../../../../../sass/breakpoints' as breakpoints;
@use '@angular/material' as mat;
@use 'sass:map';

$value-size: rem.convert(40px);
$completion-height: rem.convert(50px);
$typography: mat.define-typography-config();

.canvas {
  width: 100%;
  height: 100%;
}

.timeline {
  position: relative;
  display: flex;
  height: rem.convert(70px);
  width: 100%;

  @media screen and (max-width: breakpoints.$main-breakpoint) {
    height: rem.convert(50px);
  }

  &__line {
    flex-grow: 1;
  }

  &__name {
    display: flex;
    align-items: center;
    width: rem.convert(50px);
    font-family: monospace;
    font-size: rem.convert(20px);

    @include themes.get-foreground-color-property(color, text);
  }
}

.timeline-line {
  position: relative;
}

.timeline-line-drawing {
  width: 100%;
  height: 100%;

  &--border {
    width: rem.convert(30px);
  }

  &__line {
    @include themes.get-foreground-color-property(stroke, base);
  }
}

.timeline-arrow-drawing {
  width: 50px;

  &__line {
    @include themes.get-foreground-color-property(stroke, base);
  }

  &__polygon {
    @include themes.get-foreground-color-property(fill, base);
  }
}

.timeline-value {
  @mixin timeline-value-size($size) {
    width: $size;
    height: $size;
    top: calc(50% - #{$size} / 2);
  }

  display: block;
  position: absolute;
  left: 0;
  z-index: 4;

  @include timeline-value-size($value-size);

  @media screen and (max-width: breakpoints.$main-breakpoint) {
    @include timeline-value-size(calc(#{$value-size} / 1.5));
  }

  @mixin timeline-value-color($index, $color) {
    &--color#{$index} {
      --timeline-value-color: #{$color};
    }
  }

  @include timeline-value-color(1, map.get(mat.$red-palette, 500));
  @include timeline-value-color(2, map.get(mat.$blue-palette, 500));
  @include timeline-value-color(3, map.get(mat.$green-palette, 500));
  @include timeline-value-color(4, map.get(mat.$amber-palette, 500));
  @include timeline-value-color(5, map.get(mat.$purple-palette, 300));
  @include timeline-value-color(6, map.get(mat.$indigo-palette, 200));
  @include timeline-value-color(7, map.get(mat.$cyan-palette, 500));
  @include timeline-value-color(8, map.get(mat.$orange-palette, 500));
  @include timeline-value-color(9, map.get(mat.$lime-palette, 500));

  @media (prefers-color-scheme: dark) {
    @include timeline-value-color(1, map.get(mat.$red-palette, 800));
    @include timeline-value-color(2, map.get(mat.$blue-palette, 800));
    @include timeline-value-color(3, map.get(mat.$green-palette, 800));
    @include timeline-value-color(4, map.get(mat.$amber-palette, 800));
    @include timeline-value-color(5, map.get(mat.$purple-palette, 800));
    @include timeline-value-color(6, map.get(mat.$indigo-palette, 800));
    @include timeline-value-color(7, map.get(mat.$cyan-palette, 800));
    @include timeline-value-color(8, map.get(mat.$orange-palette, 800));
    @include timeline-value-color(9, map.get(mat.$lime-palette, 800));
  }
}

.timeline-completion {
  @mixin timeline-completion-size($val-size, $comp-height) {
    width: $val-size;
    height: $comp-height;
    top: calc(50% - #{$comp-height} / 2);
  }

  display: block;
  position: absolute;

  @include timeline-completion-size($value-size, $completion-height);

  left: 0;
  z-index: 3;

  @media screen and (max-width: breakpoints.$main-breakpoint) {
    @include timeline-completion-size(
      calc(#{$value-size} / 1.5),
      calc(#{$completion-height} / 1.5)
    );
  }
}
