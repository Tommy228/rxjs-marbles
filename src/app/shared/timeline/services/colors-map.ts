// noinspection JSUnusedGlobalSymbols
/**
 * Represents the available colors for a timeline element.
 */
export enum TimelineElementColor {
  Color1 = 'color1',
  Color2 = 'color2',
  Color3 = 'color3',
  Color4 = 'color4',
  Color5 = 'color5',
  Color6 = 'color6',
  Color7 = 'color7',
  Color8 = 'color8',
  Color9 = 'color9',
}

/**
 * Represents a color map that associates values with colors.
 */
export interface IColorsMap {
  get<T>(value: T): TimelineElementColor | undefined;

  get<T>(value: T, defaultValue: TimelineElementColor): TimelineElementColor;

  add<T>(value: T): TimelineElementColor;
}

export class ColorsMap implements IColorsMap {
  private readonly map = new Map<unknown, TimelineElementColor>();

  get<T>(value: T): TimelineElementColor | undefined;
  get<T>(value: T, defaultValue: TimelineElementColor): TimelineElementColor;
  get<T>(
    value: T,
    defaultValue?: TimelineElementColor,
  ): TimelineElementColor | undefined {
    return this.map.get(value) ?? defaultValue;
  }

  add<T>(value: T): TimelineElementColor {
    if (this.map.has(value)) {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- checked above
      return this.map.get(value)!;
    }
    const color = this.getNextColor();
    this.map.set(value, color);
    return color;
  }

  private getNextColor(): TimelineElementColor {
    const colorsCount = Object.keys(TimelineElementColor).length;
    const index = this.map.size;
    return Object.values(TimelineElementColor)[
      index % colorsCount
    ] as TimelineElementColor;
  }
}
