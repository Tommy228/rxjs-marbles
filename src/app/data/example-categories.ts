import { creationExamples } from './examples/creation-examples';
import { InjectionToken } from '@angular/core';
import { Example, ExampleFactory } from './example';
import { filteringExamples } from './examples/filtering-examples';
import { transformationExamples } from './examples/transformation-examples';
import { conditionalExamples } from './examples/conditional-examples';
import { combinationExamples } from './examples/combination-examples';
import { utilityExamples } from './examples/utility-examples';

export class ExampleCategory {
  readonly examples: readonly Example<unknown[], unknown>[] = Object.freeze(
    this.exampleFactories.map((factory) =>
      factory((x) => x as Example<unknown[], unknown>),
    ),
  );

  constructor(
    public readonly name: string,
    private readonly exampleFactories: ExampleFactory[],
  ) {}
}

export const categories: ExampleCategory[] = [
  new ExampleCategory('Combination', combinationExamples),
  new ExampleCategory('Conditional', conditionalExamples),
  new ExampleCategory('Creation', creationExamples),
  new ExampleCategory('Filtering', filteringExamples),
  new ExampleCategory('Transformations', transformationExamples),
  new ExampleCategory('Utility', utilityExamples),
];

export const EXAMPLE_CATEGORIES = new InjectionToken('EXAMPLE_CATEGORIES', {
  providedIn: 'root',
  factory: () => categories,
});
