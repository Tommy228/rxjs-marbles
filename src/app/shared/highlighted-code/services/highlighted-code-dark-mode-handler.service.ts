import { effect, inject, Injectable, Injector, Signal, untracked } from '@angular/core';
import { HighlightLoader } from 'ngx-highlightjs';
import { HighlightJsThemes } from '../provide-highlighted-code';
import { DarkModeService } from '../../dark-mode/dark-mode.service';
import { fromPromise } from 'rxjs/internal/observable/innerFrom';
import { map } from 'rxjs/operators';
import { toLazySignal } from 'ngxtension/to-lazy-signal';

/**
 * Service for handling the automatic switching of dark and light themes for highlighted code.
 * This service listens to changes in the preferred color scheme and adjusts the code highlighting theme accordingly.
 */
@Injectable({ providedIn: 'root' })
export class HighlightedCodeDarkModeHandler {
  private readonly isDarkMode = inject(DarkModeService).$isDarkMode;

  private readonly hljsLoader = inject(HighlightLoader);

  private readonly injector = inject(Injector);

  private readonly ready: Signal<boolean | undefined> = toLazySignal(
    fromPromise(this.hljsLoader.ready).pipe(map(() => true)),
    { initialValue: false }
  );

  autoSwitchDarkMode(): void {
    effect(
      () => {
        const ready = this.ready();
        if (!ready) {
          return;
        }
        const theme = this.isDarkMode()
          ? HighlightJsThemes.dark
          : HighlightJsThemes.light;
        untracked(() => this.hljsLoader.setTheme(theme));
      },
      { injector: this.injector }
    );
  }
}
