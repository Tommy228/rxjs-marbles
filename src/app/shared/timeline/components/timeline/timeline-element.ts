import { TimelineElementColor } from '../../services/colors-map';
import { TypeGuard } from '../../pipes/guard-type/guard-type-pipe';

export enum TimelineElementType {
  Value,
  Completion,
}

interface BaseTimelineElement<TType extends TimelineElementType> {
  type: TType;
  id: number;
  frame: number;
}

export interface TimelineValueElement<TValue>
  extends BaseTimelineElement<TimelineElementType.Value> {
  value: TValue;
  color: TimelineElementColor;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface TimelineCompletionElement
  extends BaseTimelineElement<TimelineElementType.Completion> {}

export type TimelineElement =
  | TimelineValueElement<unknown>
  | TimelineCompletionElement;

export const isValue: TypeGuard<
  TimelineElement,
  TimelineValueElement<unknown>
> = (element: TimelineElement): element is TimelineValueElement<unknown> =>
  element.type === TimelineElementType.Value;

export const isCompletion: TypeGuard<
  TimelineElement,
  BaseTimelineElement<TimelineElementType.Completion>
> = (element: TimelineElement): element is TimelineCompletionElement =>
  element.type === TimelineElementType.Completion;
