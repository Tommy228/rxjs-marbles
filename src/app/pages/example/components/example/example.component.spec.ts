import {
  createComponentFactory,
  mockProvider,
  Spectator,
} from '@ngneat/spectator/jest';
import { ExampleComponent } from './example.component';
import { MockComponent } from 'ng-mocks';
import { ExampleHeaderComponent } from '../example-header/example-header.component';
import { TimelineComponent } from '../../../../shared/timeline/components/timeline/timeline.component';
import { HighlightedCodeComponent } from '../../../../shared/highlighted-code/components/highlighted-code.component';
import { CurrentExampleProviderService } from '../../services/current-example-provider/current-example-provider.service';
import { Example } from '../../../../data/example';
import { map, mergeMap } from 'rxjs';
import { TestBed } from '@angular/core/testing';
import { TimelineElementColor } from '../../../../shared/timeline/services/colors-map';
import {
  TimelineElement,
  TimelineElementType,
} from '../../../../shared/timeline/components/timeline/timeline-element';
import { last } from 'lodash-es';

describe('ExampleComponent', () => {
  let spectator: Spectator<ExampleComponent>;

  const createComponent = createComponentFactory({
    component: ExampleComponent,
    declarations: [
      MockComponent(ExampleHeaderComponent),
      MockComponent(TimelineComponent),
      MockComponent(HighlightedCodeComponent),
    ],
    componentProviders: [
      mockProvider(CurrentExampleProviderService, {
        $currentExample: () =>
          ({
            name: 'mergeMap',
            label: '',
            description: '',
            linkToDocumentation: '',
            inputs: [
              [
                { t: 0, x: 'A' },
                { t: 3, x: 'B' },
                { t: 6, x: 'C' },
              ],
              [
                { t: 0, x: 1 },
                { t: 12, x: 2 },
                { t: 24, x: 3 },
                { t: 28, c: true },
              ],
            ],
            apply: ([x$, y$]) =>
              x$.pipe(mergeMap((x) => y$.pipe(map((y) => x + y)))),
          }) satisfies Example<[string, number], string>,
      }),
    ],
    shallow: true,
  });

  const getTimelines = () => spectator.queryAll(TimelineComponent);

  beforeEach(() => {
    spectator = createComponent();
    TestBed.flushEffects();
    spectator.detectChanges();
  });

  it('should create', () => {
    expect(spectator.component).toBeTruthy();
  });

  it('should set the inputs and outputs corresponding to the current example', () => {
    const timelines = getTimelines();
    expect(timelines.length).toBe(3);
    for (const timeline of timelines) {
      expect(timeline?.elements).toMatchSnapshot();
    }
  });

  it('should update the inputs and outputs when an input changes', () => {
    const firstTimeline = getTimelines()[0];
    const newTimeline: TimelineElement[] = [
      {
        color: TimelineElementColor.Color1,
        frame: 5,
        id: 0,
        type: TimelineElementType.Value,
        value: 'A',
      },
      {
        frame: 100,
        id: 1,
        type: TimelineElementType.Completion,
      },
    ];
    firstTimeline.elementsChange.emit(newTimeline);
    spectator.detectChanges();
    const timelines = getTimelines();
    expect(timelines.length).toBe(3);
    expect(timelines[0]?.elements).toEqual(newTimeline);
    expect(last(timelines)?.elements).toMatchSnapshot();
  });
});
