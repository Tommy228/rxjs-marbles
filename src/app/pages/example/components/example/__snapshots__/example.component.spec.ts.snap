// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExampleComponent should set the inputs and outputs corresponding to the current example 1`] = `
[
  {
    "color": "color1",
    "frame": 0,
    "id": 0,
    "type": 0,
    "value": "A",
  },
  {
    "color": "color2",
    "frame": 3,
    "id": 1,
    "type": 0,
    "value": "B",
  },
  {
    "color": "color3",
    "frame": 6,
    "id": 2,
    "type": 0,
    "value": "C",
  },
]
`;

exports[`ExampleComponent should set the inputs and outputs corresponding to the current example 2`] = `
[
  {
    "color": "color4",
    "frame": 0,
    "id": 0,
    "type": 0,
    "value": 1,
  },
  {
    "color": "color5",
    "frame": 12,
    "id": 1,
    "type": 0,
    "value": 2,
  },
  {
    "color": "color6",
    "frame": 24,
    "id": 2,
    "type": 0,
    "value": 3,
  },
  {
    "frame": 28,
    "id": 3,
    "type": 1,
  },
]
`;

exports[`ExampleComponent should set the inputs and outputs corresponding to the current example 3`] = `
[
  {
    "color": "color7",
    "frame": 0,
    "id": 0,
    "type": 0,
    "value": "A1",
  },
  {
    "color": "color8",
    "frame": 3,
    "id": 1,
    "type": 0,
    "value": "B1",
  },
  {
    "color": "color9",
    "frame": 6,
    "id": 2,
    "type": 0,
    "value": "C1",
  },
  {
    "color": "color1",
    "frame": 12,
    "id": 3,
    "type": 0,
    "value": "A2",
  },
  {
    "color": "color2",
    "frame": 15,
    "id": 4,
    "type": 0,
    "value": "B2",
  },
  {
    "color": "color3",
    "frame": 18,
    "id": 5,
    "type": 0,
    "value": "C2",
  },
  {
    "color": "color4",
    "frame": 24,
    "id": 6,
    "type": 0,
    "value": "A3",
  },
  {
    "color": "color5",
    "frame": 27,
    "id": 7,
    "type": 0,
    "value": "B3",
  },
  {
    "color": "color6",
    "frame": 30,
    "id": 8,
    "type": 0,
    "value": "C3",
  },
]
`;

exports[`ExampleComponent should update the inputs and outputs when an input changes 1`] = `
[
  {
    "color": "color7",
    "frame": 5,
    "id": 0,
    "type": 0,
    "value": "A1",
  },
  {
    "color": "color1",
    "frame": 17,
    "id": 1,
    "type": 0,
    "value": "A2",
  },
  {
    "color": "color4",
    "frame": 29,
    "id": 2,
    "type": 0,
    "value": "A3",
  },
  {
    "frame": 100,
    "id": 3,
    "type": 1,
  },
]
`;
